### ISSUE TITLE
Refactored jobs and files

### ISSUE THREAD

**Issue by magland** [scratchrealm/neurobass#8]
@bendichter @luiztauffer

Don't be concerned if all the files in the database have been deleted as of this afternoon. That's because I have refactored the neurobass jobs and files system. There is now a very strong provenance tracking (described below), and there are no longer .nba analysis files. The overall project layout and process for running spike sorting is smoother.

A quick description of the strong provenance system

A project consists of files and jobs. Each file is either a small (text) file with content stored in the neurobass database or is a URL pointer to a cloud file (this is the case for nwb files). In addition to a file name (within the project), each file has a unique file ID, and the content is immutable with respect to that ID. Each job has the following structure (taken from the source code)

```typescript
export type NBJob = {
    projectId: string
    workspaceId: string
    jobId: string
    userId: string
    processType: string // script, mountainsort5, kilosort3, etc.
    inputFiles: {
        name: string
        fileId: string
        fileName: string
    }[]
    inputFileIds: string[]
    inputParameters: {
        name: string
        value: any
    }[]
    outputFiles: {
        name: string
        fileName: string
        fileId?: string
    }[]
    timestampCreated: number
    computeResourceId: string
    status: 'pending' | 'queued' | 'running' | 'completed' | 'failed'
    error?: string
    processVersion?: string
    computeResourceNodeId?: string
    computeResourceNodeName?: string
    consoleOutput?: string
    timestampQueued?: number
    timestampRunning?: number
    timestampFinished?: number
    outputFileIds?: string[]
}
```

The optional "?" fields are determined at run time.

So, each job has input files and output files...

And each file that has been generated by a job has a pointer to the job that created it.

The following rules enforce the strong provenance tracking
* If a job is deleted, all of its output files are automatically deleted
* If a file is deleted, all jobs that have it as either an input or an output are deleted

So, there can be a cascading effect if a file is deleted (could trigger job deletion which would in turn trigger deletion of other files, jobs, etc).

With this system, we are guaranteed to have the full provenance history of any file in a project. In this way, the system is similar to datajoint.

I realize there are some drawbacks of this rigid system, but I thought I would try it out so we can see how we like it.

**Comment by bendichter** [comment]
This looks great. I see you have a separation between processType and processVersion. I would have gone with 

```
process: {
    type: string
    version: string
}
```

Is there an advantage to keeping these fields separate?

**Comment by bendichter** [comment]
And I guess version would be
1. If the process is a file, the hash of the file
2. If the process is a link to a dockerfile, the hash of the dockerfile

Is that more-or-less right?

**Comment by magland** [comment]
> Is there an advantage to keeping these fields separate?

Yes, one is set at creation time, and the other is set at run time.

The process is going to use some mechanism to set the version. I think it should be the version of the spike sorter (in the case of spike sorting). The hash of the docker image I think should be stored in a separate field -- some sort of runtime metadata, at the same level as the console output.
